File: objective-c-sample/SendBirdSampleUI/SendBird/SendBirdChat/View/ChannelTableViewCell.m
//
//  ChannelTableViewCell.m
//  SendBirdSampleUI
//
//  Created by SendBird Developers on 2015. 7. 29..
//  Copyright (c) 2015 SENDBIRD.COM. All rights reserved.
//

#if !__has_feature(objc_arc)
#error This file should be compiled with ARC enabled
#endif

#import "ChannelTableViewCell.h"

#define kChannelUrlFontSize 14.0
#define kChannelMembersFontSize 11.0
#define kChannelCoverRadius 19.0

@implementation ChannelTableViewCell

- (id)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    
    if (self != nil) {
        [self initViews];
    }
    return self;
}

- (void) initViews
{
    self.coverImageView = [[UIImageView alloc] init];
    [self.coverImageView setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.coverImageView.layer setCornerRadius:kChannelCoverRadius];
    [self.coverImageView setClipsToBounds:YES];
    [self.coverImageView setContentMode:UIViewContentModeScaleAspectFill];
    [self addSubview:self.coverImageView];
    
    self.channelUrlLabel = [[UILabel alloc] init];
    [self.channelUrlLabel setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.channelUrlLabel setFont:[UIFont boldSystemFontOfSize:kChannelUrlFontSize]];
    [self.channelUrlLabel setTextColor:UIColorFromRGB(0x414858)];
    [self addSubview:self.channelUrlLabel];
    
    self.memberCountLabel = [[UILabel alloc] init];
    [self.memberCountLabel setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.memberCountLabel setFont:[UIFont systemFontOfSize:kChannelMembersFontSize]];
    [self.memberCountLabel setTextColor:UIColorFromRGB(0xa6b0ba)];
    [self addSubview:self.memberCountLabel];
    
    self.checkImageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"_icon_check"]];
    [self.checkImageView setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.checkImageView setHidden:YES];
    [self addSubview:self.checkImageView];
    
    self.bottomLineView = [[UIView alloc] init];
    [self.bottomLineView setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.bottomLineView setBackgroundColor:UIColorFromRGB(0xd9d9d9)];
    [self addSubview:self.bottomLineView];
    
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.coverImageView
                                                     attribute:NSLayoutAttributeCenterY
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeCenterY
                                                    multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.coverImageView
                                                     attribute:NSLayoutAttributeLeading
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeLeading
                                                    multiplier:1 constant:15]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.coverImageView
                                                     attribute:NSLayoutAttributeWidth
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:38]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.coverImageView
                                                     attribute:NSLayoutAttributeHeight
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:38]];
    
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.channelUrlLabel
                                                     attribute:NSLayoutAttributeTop
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeTop
                                                    multiplier:1 constant:6]];
    
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.channelUrlLabel
                                                     attribute:NSLayoutAttributeLeading
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.coverImageView
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:15]];
    
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.channelUrlLabel
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:-46]];
    
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.memberCountLabel
                                                     attribute:NSLayoutAttributeBottom
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeBottom
                                                    multiplier:1 constant:-6]];
    
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.memberCountLabel
                                                     attribute:NSLayoutAttributeLeading
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.coverImageView
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:15]];
    
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.memberCountLabel
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:-46]];
    
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.checkImageView
                                                     attribute:NSLayoutAttributeCenterY
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeCenterY
                                                    multiplier:1 constant:0]];
    
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.checkImageView
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:-15]];
    
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.checkImageView
                                                     attribute:NSLayoutAttributeWidth
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:22]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.checkImageView
                                                     attribute:NSLayoutAttributeHeight
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:22]];
    
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.bottomLineView
                                                     attribute:NSLayoutAttributeBottom
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeBottom
                                                    multiplier:1 constant:0]];
    
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.bottomLineView
                                                     attribute:NSLayoutAttributeLeading
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.coverImageView
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:15]];
    
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.bottomLineView
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:0]];
    
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.bottomLineView
                                                     attribute:NSLayoutAttributeHeight
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:1]];
}

- (void) setModel:(SendBirdChannel *)model
{
    if (model == nil) {
        return;
    }
    
    [self.channelUrlLabel setText:[NSString stringWithFormat:@"#%@",[SendBirdUtils getChannelNameFromUrl:[model url]]]];
    if ([model memberCount] <= 1) {
        [self.memberCountLabel setText:[NSString stringWithFormat:@"%d MEMBER", [model memberCount]]];
    }
    else {
        [self.memberCountLabel setText:[NSString stringWithFormat:@"%d MEMBERS", [model memberCount]]];
    }
    
    if ([[SendBird getCurrentChannel] channelId] == [model channelId]) {
        [self setBackgroundColor:UIColorFromRGB(0xffffe2)];
        [self.checkImageView setHidden:NO];
    }
    else {
        [self setBackgroundColor:[UIColor clearColor]];
        [self.checkImageView setHidden:YES];
    }

    [SendBirdUtils loadImage:[model coverUrl] imageView:self.coverImageView width:38 height:38];
}

@end

File: swift-sample-v2/SendBirdSampleUI/SendBird/SendBirdChat/View/ChannelTableViewCell.swift
//
//  ChannelTableViewCell.swift
//  SendBirdSampleUI
//
//  Created by Jed Kyung on 2/2/16.
//  Copyright © 2016 SENDBIRD.COM. All rights reserved.
//

import UIKit
import SendBirdSDK

let kChannelUrlFontSize: CGFloat = 14.0
let kChannelMembersFontSize: CGFloat = 11.0
let kChannelCoverRadius: CGFloat = 19.0

class ChannelTableViewCell: UITableViewCell {
    var channelUrlLabel: UILabel?
    var memberCountLabel: UILabel?
    var coverImageView: UIImageView?
    var checkImageView: UIImageView?
    var bottomLineView: UIView?

    override func awakeFromNib() {
        super.awakeFromNib()
        // Initialization code
    }

    override func setSelected(selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)

        // Configure the view for the selected state
    }

    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        initView()
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func initView() {
        self.coverImageView = UIImageView()
        self.coverImageView?.translatesAutoresizingMaskIntoConstraints = false
        self.coverImageView?.layer.cornerRadius = kChannelCoverRadius
        self.coverImageView?.clipsToBounds = true
        self.coverImageView?.contentMode = UIViewContentMode.ScaleAspectFill
        self.addSubview(self.coverImageView!)
        
        self.channelUrlLabel = UILabel()
        self.channelUrlLabel?.translatesAutoresizingMaskIntoConstraints = false
        self.channelUrlLabel?.font = UIFont.boldSystemFontOfSize(kChannelUrlFontSize)
        self.channelUrlLabel?.textColor = SendBirdUtils.UIColorFromRGB(0x414858)
        self.addSubview(self.channelUrlLabel!)
        
        self.memberCountLabel = UILabel()
        self.memberCountLabel?.translatesAutoresizingMaskIntoConstraints = false
        self.memberCountLabel?.font = UIFont.systemFontOfSize(kChannelMembersFontSize)
        self.memberCountLabel?.textColor = SendBirdUtils.UIColorFromRGB(0xa6b0ba)
        self.addSubview(self.memberCountLabel!)
        
        self.checkImageView = UIImageView.init(image: UIImage.init(named: "_icon_check"))
        self.checkImageView?.translatesAutoresizingMaskIntoConstraints = false
        self.checkImageView?.hidden = true
        self.addSubview(self.checkImageView!)
        
        self.bottomLineView = UIView()
        self.bottomLineView?.translatesAutoresizingMaskIntoConstraints = false
        self.bottomLineView?.backgroundColor = SendBirdUtils.UIColorFromRGB(0xd9d9d9)
        self.addSubview(self.bottomLineView!)
        
        // Cover Image View
        self.addConstraint(NSLayoutConstraint.init(item: self.coverImageView!, attribute: NSLayoutAttribute.CenterY, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.CenterY, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.coverImageView!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Leading, multiplier: 1, constant: 15))
        self.addConstraint(NSLayoutConstraint.init(item: self.coverImageView!, attribute: NSLayoutAttribute.Width, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: 38))
        self.addConstraint(NSLayoutConstraint.init(item: self.coverImageView!, attribute: NSLayoutAttribute.Height, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: 38))

        // Channel URL Label
        self.addConstraint(NSLayoutConstraint.init(item: self.channelUrlLabel!, attribute: NSLayoutAttribute.Top, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Top, multiplier: 1, constant: 6))
        self.addConstraint(NSLayoutConstraint.init(item: self.channelUrlLabel!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self.coverImageView, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: 15))
        self.addConstraint(NSLayoutConstraint.init(item: self.channelUrlLabel!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: -46))
        
        // Member Count Label
        self.addConstraint(NSLayoutConstraint.init(item: self.memberCountLabel!, attribute: NSLayoutAttribute.Bottom, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Bottom, multiplier: 1, constant: -6))
        self.addConstraint(NSLayoutConstraint.init(item: self.memberCountLabel!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self.coverImageView, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: 15))
        self.addConstraint(NSLayoutConstraint.init(item: self.memberCountLabel!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: -46))
        
        // Check Image View
        self.addConstraint(NSLayoutConstraint.init(item: self.checkImageView!, attribute: NSLayoutAttribute.CenterY, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.CenterY, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.checkImageView!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: -15))
        self.addConstraint(NSLayoutConstraint.init(item: self.checkImageView!, attribute: NSLayoutAttribute.Width, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: 22))
        self.addConstraint(NSLayoutConstraint.init(item: self.checkImageView!, attribute: NSLayoutAttribute.Height, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: 22))

        // Bottom Line View
        self.addConstraint(NSLayoutConstraint.init(item: self.bottomLineView!, attribute: NSLayoutAttribute.Bottom, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Bottom, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.bottomLineView!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self.coverImageView, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: 15))
        self.addConstraint(NSLayoutConstraint.init(item: self.bottomLineView!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.bottomLineView!, attribute: NSLayoutAttribute.Height, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: 1))

    }
    
    func setModel(model: SendBirdChannel?) {
        if model == nil {
            return
        }
        
        self.channelUrlLabel?.text = String.init(format: "#%@", SendBirdUtils.getChannelNameFromUrl((model?.url)!))
        if model?.memberCount <= 1 {
            self.memberCountLabel?.text = String.init(format: "%d MEMBER", (model?.memberCount)!)
        }
        else {
            self.memberCountLabel?.text = String.init(format: "%d MEMBERS", (model?.memberCount)!)
        }
        
        if SendBird.getCurrentChannel() != nil && SendBird.getCurrentChannel().channelId == model?.channelId {
            backgroundColor = SendBirdUtils.UIColorFromRGB(0xffffe2)
            self.checkImageView?.hidden = false
        }
        else {
            backgroundColor = UIColor.clearColor()
            self.checkImageView?.hidden = true
        }
        
        SendBirdUtils.loadImage((model?.coverUrl)!, imageView: self.coverImageView!, width: 38, height: 38)
    }
}
